### DISCUSSION TITLE
UWB system

### DISCUSSION THREAD

**Discussion by CalvinKoung** [bitcraze/discussions#1006] [Q&A]
Hi, I want to conduct research on the UWB algorithm. For instance, I want to try to see if Crazyflie can select the best 3 or 4 anchors for indoor positioning in an environment with many anchors, and filter other redundant signals to increase the accuracy of positioning. Does anyone can suggest me which file I can read or open source I can try to modify. Thanks.

**Comment by knmcguire** [comment]
Hi! That sounds doable but let's start with giving you a list of read materials for you to get started on:

- Landing page of the loco positioning system which is the hub that links to a lot of other handy documentation: https://www.bitcraze.io/documentation/system/positioning/loco-positioning-system/
- The repository more detailed documentation of the loco positioning system: https://www.bitcraze.io/documentation/repository/lps-node-firmware/master/
- The lo[co positioning measurement models ](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/loco-positioning-system/)used in [the extended kalman filter](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/sensor-to-control/state_estimators/#extended-kalman-filter)
- The locodeck driver in the crazyflie firmware: https://github.com/bitcraze/crazyflie-firmware/blob/master/src/deck/drivers/src/locodeck.c. The seperate code for the TWR, TDOA2 and TDOA3 code is found in the same folder.
- This blogpost about the updated TDOA long range filter will also be interesting for you: https://www.bitcraze.io/2023/03/new-loco-tdoa-outlier-filter/

Hope this can get you started in the right directions! Let us know if you still have any questions about it still.

**Reply by CalvinKoung** [reply]
Ok, thanks your help. I will do some research.

**Reply by CalvinKoung** [reply]
Hi, I have a question. If I want to create a new algorithm for UWB system. I need to upload in UWB firmware is this correct ? thanks

**Reply by knmcguire** [reply]
If you want onboard estimation in a new way, yes you will need to implement it yourself onboard of the crazyflie firmware. 

You should be able to try things out offboard though. There is this very old example from the early days from ultrawide band when the estimation was done on an external computer https://github.com/bitcraze/lps-ros

**Reply by CalvinKoung** [reply]
Ok, I got it. But where can I find the original UWB code in the crazyflie. I want to understand the original code and then try to modify it. Can I do this ?

**Reply by krichardsson** [reply]
You can find the code in the [Crazfylie firmware repository](https://github.com/bitcraze/crazyflie-firmware) 
Start looking at the files named "lpsXXX" in the [src/deck/driver/src directory](https://github.com/bitcraze/crazyflie-firmware/tree/master/src/deck/drivers/src)
The TdoA favors also use the tdoa engine that you can find in the `src/modules` tree as well as in the `src/utils/src/tdoa` directories
